Lecture_2/Exercise_3/list.c

```diff
@@ -0,0 +1,147 @@
+#include "list.h"
+
+list* ListAdd(list **root_ptr, void *content_ptr)
+{
+        list *list_ptr = *root_ptr;
+
+	list *curr_item = (list*)malloc(sizeof(list));
+	curr_item->content = content_ptr;
+
+	if (list_ptr == NULL) {
+		curr_item->prev = NULL;
+		curr_item->next = NULL;
+
+		*root_ptr = curr_item;	// update root by first
+	} else {
+		list_ptr = ListGetLast(list_ptr);
+
+		curr_item->prev = list_ptr;
+		curr_item->next = list_ptr->next;
+
+		list_ptr->next = curr_item;
+	}
+	return curr_item;
+}
+
+list* ListAddBefore(list **root_ptr, uint32_t before_id, void *content_ptr)
+{
+	list *list_ptr = ListGet(*root_ptr, before_id);
+
+	if (list_ptr != NULL) {
+		list *curr_item = (list*)malloc(sizeof(list));
+		curr_item->content = content_ptr;
+
+		if (before_id == 0) {
+			*root_ptr = curr_item;		// set item as first
+		} else {
+			list_ptr->prev->next = curr_item;
+		}
+		curr_item->prev = list_ptr->prev;
+		curr_item->next = list_ptr;
+		list_ptr->prev = curr_item;
+
+		return curr_item;
+	}
+	return NULL;		// unreachable before_id
+}
+
+uint32_t ListGetLength(list *list_ptr)
+{
+	uint32_t counter = 0;
+
+	do {
+		counter++;
+		list_ptr = list_ptr->next;
+	} while (list_ptr != NULL);
+
+	return counter;
+}
+
+list* ListGetLast(list *list_ptr)
+{
+	while (list_ptr->next != NULL) {
+		list_ptr = list_ptr->next;
+	}
+	return list_ptr;
+}
+
+list* ListGet(list *list_ptr, uint32_t index)
+{
+	uint32_t i = 0;
+	while(list_ptr != NULL) {
+		if (i == index) {
+			return list_ptr;
+		} else {
+			list_ptr = list_ptr->next;
+			i++;
+		}
+	}
+	return NULL;		// unexpected end of list
+}
+
+void* ListIs(list *list_ptr, uint32_t index)
```

Comment:
The idea of Listls function was to print all values from list (just like an ls command in linux terminal), not to read a single value.

But it's ok as it was not very clear, and you've implemented the "print all" functionality within the menu of main.c.